crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+                      uint8_t *tmp_ctext,
+                      const scalar *m,
+                      const scalar *s,
+                      const scalar *t,
+                      const uint8_t rho[ML_KEM_RANDOM_BYTES],
+                      const uint8_t pkhash[ML_KEM_PKHASH_BYTES],
+                      const uint8_t z[ML_KEM_RANDOM_BYTES],
+                      scalar *tmp_y,
+                      scalar *tmp_e1,
+                      scalar *tmp_u,
+                      vinfo_t vinfo,
+                      mctx *ctx)
+{
+    uint8_t decrypted[ML_KEM_SHARED_SECRET_BYTES + ML_KEM_PKHASH_BYTES];
+    uint8_t Kr[ML_KEM_SHARED_SECRET_BYTES + ML_KEM_RANDOM_BYTES];
+    uint8_t failure_key[ML_KEM_RANDOM_BYTES];
+    uint8_t *r = Kr + ML_KEM_SHARED_SECRET_BYTES;
+    uint8_t mask;
+    int i;
+
+# if ML_KEM_SHARED_SECRET_BYTES != ML_KEM_RANDOM_BYTES
+#  error "Invalid unequal lengths of ML-KEM shared secret and random inputs"
+# endif
+
+    /*
+     * If our KDF is unavailable, fail early! Otherwise, keep going ignoring
+     * any further errors, returning success, and whatever we got for a shared
+     * secret.  The decrypt_cpa() function is just arithmetic on secret data,
+     * so should not be subject to failure that makes its output predictable.
+     *
+     * We guard against "should never happen" catastrophic failure of the
+     * "pure" function |hash_g| by overwriting the shared secret with the
+     * content of the failure key and returning early, if nevertheless hash_g
+     * fails.  This is not constant-time, but a failure of |hash_g| already
+     * implies loss of side-channel resistance.
+     *
+     * The same action is taken, if also |encrypt_cpa| should catastrophically
+     * fail, due to failure of the |PRF| underlyign the CBD functions.
+     */
+    if (!kdf(failure_key, z, ctext, vinfo->ctext_bytes, ctx))
+        return 0;
+    decrypt_cpa(decrypted, ctext, s, tmp_u, vinfo);
+    memcpy(decrypted + ML_KEM_SHARED_SECRET_BYTES, pkhash, ML_KEM_PKHASH_BYTES);
+    if (!hash_g(Kr, decrypted, sizeof(decrypted), ctx)
+        || !encrypt_cpa(tmp_ctext, decrypted, m, t, r, tmp_y, tmp_e1, tmp_u,
+                        vinfo, ctx)) {
+        memcpy(shared_secret, failure_key, ML_KEM_SHARED_SECRET_BYTES);
+        return 1;
+    }
+    mask = constant_time_eq_int_8(0,
+        CRYPTO_memcmp(ctext, tmp_ctext, vinfo->ctext_bytes));
+    for (i = 0; i < ML_KEM_SHARED_SECRET_BYTES; i++)
+        shared_secret[i] = constant_time_select_8(mask, Kr[i], failure_key[i]);
+    return 1;
+}
+
+/*
+ * -----
+ *
+ * Perform ML_KEM ops for a specific parameter set, the public and private key
+ * pointers are (void *), in that order, when either is accepted in lieu of a
+ * public key of the appropriate type.  Parameters with an implicit fixed
+ * length in each variant have an explicit checked additional length argument
+ * here.
+ */
+# define ml_kem_bname(bits, b)  ossl_ml_kem_##bits##_##b
```

Comment:
Move defines to top of file and uppercase them

Reply:
Gone in #26172